File: ruby/lib/mudbase/capacity.rb
# frozen_string_literal: true

require_relative "bookings_repo"
require_relative "activity_repo"

module Mudbase
  # Ports the reference Next.js app's `src/lib/capacity.ts` (`reconcileEventCapacity`) verbatim -
  # see plan/build-plan.md "Capacity-Race Handling Approach" for the full rationale. Mudbase (a
  # generic-CRUD BaaS) has no cross-document transactions or atomic counters, so a plain "count
  # confirmed, then create" is inherently racy: two simultaneous booking requests can both read
  # the same pre-write count and both decide "there's room". This module narrows that race window
  # by re-deriving truth from a fresh read after every booking create/cancel: creation-order
  # priority (the first `capacity` bookings, oldest first, among confirmed+waitlisted) are the
  # ones entitled to a seat; any booking whose current status disagrees with that derivation gets
  # corrected, with a matching `activity` entry logged for the correction.
  #
  # Deliberately excludes `"checked_in"` bookings from the capacity count (same reading of the
  # task's literal spec as the reference app) - capacity is defined in terms of `"confirmed"`
  # bookings specifically, and running this after check-in would incorrectly free an
  # already-seated attendee's slot for someone else on the waitlist.
  module Capacity
    def self.reconcile!(access_token:, event_id:, capacity:)
      live = Mudbase::BookingsRepo.live_for_event(access_token: access_token, event_id: event_id)

      live.each_with_index do |booking, index|
        should_be_confirmed = index < capacity

        if should_be_confirmed && booking[:status] != "confirmed"
          correct!(access_token, booking, "confirmed", "booking_promoted")
        elsif !should_be_confirmed && booking[:status] != "waitlisted"
          correct!(access_token, booking, "waitlisted", "booking_waitlisted")
        end
      end
    end

    def self.correct!(access_token, booking, new_status, action)
      Mudbase::BookingsRepo.update!(access_token: access_token, id: booking[:_id], attributes: { status: new_status })
      Mudbase::ActivityRepo.create!(
        access_token: access_token,
        attributes: {
          eventId: booking[:eventId],
          actorId: booking[:userId],
          actorName: booking[:userName],
          action: action,
        },
      )
    end
    private_class_method :correct!
  end
end

File: mobile-flutter/lib/core/capacity.dart
import '../data/repositories/activity_repository.dart';
import '../data/repositories/booking_repository.dart';
import '../models/activity_entry.dart';
import '../models/booking.dart';

/// Re-derives which bookings for an event should hold a confirmed seat
/// versus sit on the waitlist, and patches only the ones whose current
/// status disagrees with that derivation. Ports `reconcileEventCapacity` in
/// the reference web app's `src/lib/capacity.ts` verbatim - see that file
/// for the full rationale; the summary:
///
/// Mudbase (a generic-CRUD BaaS) has no cross-document transactions or
/// atomic counters, so a plain "count confirmed, then create" is inherently
/// racy: two simultaneous booking requests can both read the same pre-write
/// count and both decide "there's room". This function narrows that race
/// window by re-deriving truth from a fresh read (creation-order priority:
/// the first `capacity` bookings, oldest first, among confirmed+waitlisted,
/// are the ones entitled to a seat) and correcting any booking that
/// disagrees - whether that means demoting an overshoot back to waitlisted,
/// or promoting the earliest waitlisted booking once a cancellation frees a
/// seat.
///
/// Deliberately excludes `checkedIn` bookings from the capacity count (see
/// `plan/build-plan.md` "Capacity-Race Handling Approach") - the task's spec
/// defines capacity in terms of confirmed bookings specifically, and running
/// this after check-in would incorrectly free an already-seated attendee's
/// slot for someone else.
Future<void> reconcileEventCapacity({
  required String token,
  required BookingRepository bookingRepository,
  required ActivityRepository activityRepository,
  required String eventId,
  required int capacity,
}) async {
  final results = await Future.wait([
    bookingRepository.listForEvent(
      token: token,
      eventId: eventId,
      status: 'confirmed',
      sort: 'createdAt',
      limit: BookingRepository.reconcileFetchLimit,
    ),
    bookingRepository.listForEvent(
      token: token,
      eventId: eventId,
      status: 'waitlisted',
      sort: 'createdAt',
      limit: BookingRepository.reconcileFetchLimit,
    ),
  ]);

  final live = deriveReconciliationOrder([...results[0], ...results[1]]);

  final corrections = <Future<void>>[];
  for (var index = 0; index < live.length; index++) {
    final booking = live[index];
    final shouldBeConfirmed = index < capacity;

    if (shouldBeConfirmed && booking.status != BookingStatus.confirmed) {
      corrections.add(
        _promote(
          token,
          bookingRepository,
          activityRepository,
          eventId,
          booking,
        ),
      );
    } else if (!shouldBeConfirmed &&
        booking.status != BookingStatus.waitlisted) {
      corrections.add(
        _demote(token, bookingRepository, activityRepository, eventId, booking),
      );
    }
  }

  await Future.wait(corrections);
}

/// The pure derivation half of reconciliation, split out from the I/O above
/// specifically so it is unit-testable with plain `dart test` and no network
/// dependency: sorts confirmed+waitlisted bookings by creation order
/// (oldest first), which is the priority order reconciliation applies.
List<Booking> deriveReconciliationOrder(List<Booking> confirmedAndWaitlisted) {
  final sorted = [...confirmedAndWaitlisted]
    ..sort((a, b) => a.createdAt.compareTo(b.createdAt));
  return sorted;
}

Future<void> _promote(
  String token,
  BookingRepository bookingRepository,
  ActivityRepository activityRepository,
  String eventId,
  Booking booking,
) async {
  await bookingRepository.updateStatus(
    token: token,
    bookingId: booking.id,
    status: BookingStatus.confirmed,
  );
  await activityRepository.log(
    token: token,
    eventId: eventId,
    actorId: booking.userId,
    actorName: booking.userName,
    action: ActivityAction.bookingPromoted,
  );
}

Future<void> _demote(
  String token,
  BookingRepository bookingRepository,
  ActivityRepository activityRepository,
  String eventId,
  Booking booking,
) async {
  await bookingRepository.updateStatus(
    token: token,
    bookingId: booking.id,
    status: BookingStatus.waitlisted,
  );
  await activityRepository.log(
    token: token,
    eventId: eventId,
    actorId: booking.userId,
    actorName: booking.userName,
    action: ActivityAction.bookingWaitlisted,
  );
}
